### PR TITLE
ci: `cargo-semver-checks`

### LINKED ISSUE #1875: Check for API breaking changes in CI
As we move towards a 1.0 release it would be nice to check CI that we don't make breaking changes. Consider using 
- https://github.com/Enselic/cargo-public-api 
- https://github.com/obi1kenobi/cargo-semver-checks

### PR DESCRIPTION
**Pull request by storopoli**
- Removes `check-api` (`cargo-public-api`)
- Adds `cargo-semver-checks`

## Note to Reviewers

There is a new script `contrib/check-semver.sh` that checks for semver breaks against `master`.
If it detects a breaking change it will create the `.semver-break` file.
If this file exists then we will add an `API break` label and a big comment to the PR saying:

> 🚨 API BREAKING CHANGE DETECTED

It reproduces the current behavior of `cargo-public-api`:

- `bitcoin`:
  - `all-features`
  - `no-default-features`
- `base58ck`:
  - `all-features`
  - `no-default-features`
- `bitcoin_hashes`:
  - `no-default-features`
  - `features alloc`
- `bitcoin-units`:
  - `no-default-features`
  - `features alloc`
- `bitcoin-io`:
  - `no-default-features`
  - `features alloc`

Closes #1875.
Supersedes #2912.

## Context

Our current test to check API breaks using `cargo public-api` is not "automated" per se.
It needs contributors to keep tabs on text files with function signatures and other things and forces reviewers to be the _de facto_ API break detector tool.

The idea is to use the well-acclaimed and maintained `cargo semver-checks` that will do this automatically in CI.

### CODE REVIEW

**Comment by storopoli** [comment]
@obi1kenobi, I think that for the two items in the TODO list above we cannot do it with the `obi1kenobi/cargo-semver-checks-action` action right?

I'd probably need to run `cargo semver-checks` with some customizations and extract the information,
then pipe it to `actions/labeler` right?

**Comment by obi1kenobi** [comment]
To not fail the workflow, you could use GitHub Actions' `continue-on-error` argument: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error

The labeling workflow is not possible with the current action, since it always uses the version on crates.io as baseline, not the repo's main branch. That means that after you merge the first breaking change, all subsequent workflow runs until the next release will re-discover that breaking change since they will be comparing "PR to crates.io" not "PR to target branch." I want this to get solved in the action, but that's blocked [on finding more funding](https://predr.ag/blog/four-challenges-cargo-semver-checks-has-yet-to-tackle/#existential-threat-sustainable-project-funding) right now. If you come up with a workflow you like, I'd love to see it and possibly even accept a PR to the action.

One more caveat: support for nightly Rust versions in `cargo-semver-checks` is best-effort and not guaranteed. The current beta and a range of stable versions are supported, as indicated in the release notes. This is not to say you can't use it with nightly, you just have to be careful when you bump nightly versions. I don't offer any SLA for nightly support without a formal support contract: nightly Rust can ship a new rustdoc JSON format at any time without warning and it's infeasible for me to offer any kind of SLA or guarantees for free.

**Comment by apoelstra** [comment]
Sounds good. We'll dig into what's needed to use a different baseline. Presumably we won't be able to use the action, and instead will need to hack up some shell scripts to do it for us. This shouldn't be too bad. We have some experience with our current label-bot doing things like this.

**Comment by obi1kenobi** [comment]
Thank you, I really appreciate it!

`cargo-semver-checks` supports specifying a different baseline explicitly, so I imagine the work would be mostly around plumbing that through the action and figuring out how to connect the CLI to the labeling bot.

I've also wanted to make the action annotate the code: post notes or comments on _specific lines_ where breakage happened. I'll describe the stumbling blocks I ran into, in case that's helpful:
- GitHub's existing annotation support in actions only allows annotating the "new" code. This is fine for regular linters, but not great here: `cargo-semver-checks` may need to point to a function/struct/enum etc. that used to exist but has been deleted, which we cannot do within that GitHub capability.
- Posting comments and setting labels seems to require elevated privileges compared to what PRs usually have. We don't want to run the action on the `pull_request_target` trigger, because a malicious PR coming from a fork could run an arbitrary `build.rs` script as part of `cargo-semver-checks` and steal credentials etc.

**Comment by Kixunil** [comment]
> * Posting comments and setting labels seems to require elevated privileges compared to what PRs usually have.

What you need to do is somehow export the information to other actions and have a flag to not fail. Then other labeling action can take that and use it. Though to be strictly secure it'd need to be another job I think. (Not a big problem I guess.)

**Comment by storopoli** [comment]
Apparently `cargo semver-checks` hardcodes stuff to the `RUSTDOCFLAGS` (check https://github.com/obi1kenobi/cargo-semver-checks/blob/5108eeeffaf424e8d1a0019534414e7f11276a66/src/rustdoc_cmd.rs#L111-L113).

And we need different stuff (see https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/blob/3494ceec521cbc2cbe7e4b0d3fb9219a2164f57d/ci/run_task.sh#L277)

So, I will try to generate the JSON files manually and call `cargo semver-checks` to match only on these JSONs.

File: contrib/check-semver.sh
Comment (storopoli) [inline contrib/check-semver.sh:57]: @obi1kenobi how do I save the `cargo semver-checks` output to a file?

**Comment by Kixunil** [comment]
> * You don't want to semver-check with `--cfg docsrs` set, because that isn't the API that users of the crate actually use.

It really shouldn't. We only use it to enable the unstable feature that says which items are activated by which features.

> * adding a private type or even a private `use` import item can cause a major breaking change

OMG :man_facepalming: :man_facepalming: :man_facepalming:

File: contrib/check-semver.sh
Comment (obi1kenobi) [reply contrib/check-semver.sh:57]: Append `2>&1` to the command. Right now you are sending stdout to the file, but stderr still gets printed normally. If you do `>file.txt 2>&1` then that sends both stdout and stderr to the file. The order of these two commands matters, so make sure `2>&1` is second.

Comment (obi1kenobi) [reply contrib/check-semver.sh:57]: We don't have a structured output format for the cargo-semver-checks findings, but I'm open to adding one. We just have to work out a good name for the CLI flag, and how to version it and keep it stable across versions.

Comment (storopoli) [reply contrib/check-semver.sh:57]: For now I just want to `grep` a fail in the `-v` (verbose) output and if positive, add a label to the PR (and maybe a comment as well?).

File: contrib/check-semver.sh
Comment (storopoli) [inline contrib/check-semver.sh:13]: Note to myself: this does not work... Why? I added the `fetch-depth: 0`

```yml
      - name: "Checkout repo"
        uses: actions/checkout@v4
        with:
          # Number of commits to fetch. 0 indicates all history for all branches and tags.
          fetch-depth: 0 # we need full history for cargo semver-checks
```

Ref: https://github.com/actions/checkout

Comment (obi1kenobi) [reply contrib/check-semver.sh:13]: Is it possibly related to the fact it's running inside a forked repo, or something? Not sure... 🤔

Comment (apoelstra) [reply contrib/check-semver.sh:13]: @storopoli github provides some magic variables that will have git commit hashes for the base and head branches. You want to use those rather than hardcoding `master`, which aside from not working, is not always correct because some PRs are not against master.

Comment (storopoli) [reply contrib/check-semver.sh]: Done! Thanks :)

**Comment by storopoli** [comment]
Almost there. I am still running into issues with all the other crates in https://github.com/rust-bitcoin/rust-bitcoin/blob/b8861c2d02c3cc0dd1def3682345e53b15aa0bee/contrib/check-semver.sh#L29-L33

except `bitcoin`.

Getting:

```
warning: /home/runner/work/rust-bitcoin/rust-bitcoin/bitcoin/Cargo.toml: version requirement `0.106.0+26` for dependency `bitcoinconsensus` includes semver metadata which will be ignored, removing the metadata is recommended to avoid confusion
error: cannot specify features for packages outside of workspace
```

File: contrib/check-semver.sh
Comment (Kixunil) [inline contrib/check-semver.sh:49]: You should generate these programmatically.

Comment (storopoli) [reply contrib/check-semver.sh:49]: I don't follow. Programmatically means "in a way that follows a plan or uses a particular method".
That can be anything. Can you be more specific?

Comment (apoelstra) [reply contrib/check-semver.sh:49]: I believe he means "extract the names of the crates from the top-level Cargo.toml file".

**Comment by apoelstra** [comment]
> warning: /home/runner/work/rust-bitcoin/rust-bitcoin/bitcoin/Cargo.toml: version requirement `0.106.0+26` for dependency `bitcoinconsensus` includes semver metadata which will be ignored, removing the metadata is recommended to avoid confusion
> error: cannot specify features for packages outside of workspace

I wouldn't worry about this warning. I wish we could silence it. It's true that the `+26` provides no additional information to Cargo.toml but it **does** signal to the user that this is the version of bitcoinconsensus that has Bitcoin 26. I guess we could truncate the version and then move the full version to a comment, explaining why Cargo makes us do that, but come on.

**Comment by Kixunil** [comment]
> I guess we could truncate the version and then move the full version to a comment, explaining why Cargo makes us do that, but come on.

I think the warning is correct. The metadata in the requirement field is noop, so putting it there is equivalent to writing a comment which doesn't look like a comment, so it's confusing. We can achieve the same by writing a comment and then whoever reads it know it's just a comment and they can learn the true version by running `cargo tree`.

File: .github/workflows/rust.yml
Comment (storopoli) [inline .github/workflows/rust.yml:322]: Depending on the threat model, we can speed this up with:

```suggestion
      - name: "Install cargo-binstall"
        uses: cargo-bins/cargo-binstall@main
      - name: "Binstall cargo-semver-checks"
        run: cargo binstall cargo-semver-checks --no-confirm
```

Comment (Kixunil) [reply .github/workflows/rust.yml:322]: I think it's fine. We're not checking the signatures/hashes of the source either.

**Comment by storopoli** [comment]
Finally! it gives no error! I will try to make a PR on top of this with a breaking change. Let's see!

File: contrib/check-semver.sh
Comment (obi1kenobi) [inline contrib/check-semver.sh:74]: Please make sure to use all the flags `cargo-semver-checks` uses to generate rustdoc JSON, or else you'll get unexpected behavior from the tool like false-positives, false-negatives, and crashes. Those will be hell to debug because they won't reproduce when you run `cargo-semver-checks` directly.

https://github.com/obi1kenobi/cargo-semver-checks/blob/5108eeeffaf424e8d1a0019534414e7f11276a66/src/rustdoc_cmd.rs#L113

Comment (storopoli) [reply contrib/check-semver.sh:74]: Yeah check https://github.com/rust-bitcoin/rust-bitcoin/blob/264051814b858a78b96d78e45bf468b9dba7fae7/contrib/check-semver.sh#L11

Comment (obi1kenobi) [reply contrib/check-semver.sh:74]: Ah I missed that, my bad!

**Comment by storopoli** [comment]
Ok this is ready to review.
Thanks @obi1kenobi for the amazing support. It would definitely take me at least 4 days to figure everything on my own.

You can see it in action in this PR https://github.com/storopoli/rust-bitcoin/pull/5.
It detected the breaking change, added `API Break` label and commented in the PR.

**Comment by apoelstra** [comment]
Super cool!! It would be nice to comment with more details. And I think maybe one red-siren emoji and non-bold text would be enough to get our attention :).

**Comment by storopoli** [comment]
> Super cool!! It would be nice to comment with more details. And I think maybe one red-siren emoji and non-bold text would be enough to get our attention :).

Ok done everything. I was extra descriptive in the comments. We can always remove some if necessary.

**Review by tcharding — DISMISSED** [review]
Awesome work man, I can see you put in a whole lot of effort! I have a few minor nits but I think we should merge as is and improve in follow ups if we want to, `bash` _always_ has a few nits :)

To be explicit, and for other reviewers these are the issues I see that I think are ok to merge:
- code duplication
- code comments
- I _think_ `bitcoin_hashes` might end up being `bitcoin-hashes` in some file names

Thanks again @storopoli, props on this one.

**Comment by tcharding** [comment]
This PR should probably also remove the `api/` directory.

**Comment by storopoli** [comment]
> This PR should probably also remove the `api/` directory.

Yes, good catch. Did this in the proper commit d9567b09.
This, by now, should be ready to review and merge.

**Review by apoelstra — APPROVED** [review]
ACK bc2b980b10c94e2872ca1480eb11c7e0cb16b22e Looks great -- definitely want to circle back to improve the comment it posts, try to clean up some repeated code, and possibly add a larger feature matrix

File: contrib/check-semver.sh
Comment (Kixunil) [inline contrib/check-semver.sh:13]: I always read this as "euro" :D

File: contrib/check-semver.sh
Comment (Kixunil) [inline contrib/check-semver.sh:102]: Oh, wait, does it catch public deps anyway or not?

File: contrib/check-semver.sh
Comment (Kixunil) [inline contrib/check-semver.sh:114]: There's a less brittle hack for that: append `|| true` at the end of the command.

File: contrib/check-semver.sh
Comment (Kixunil) [inline contrib/check-semver.sh:183]: Now I really don't understand why it's done here and not after the check.

**Review by Kixunil — APPROVED** [review]
ACK eeae225cfca149d669b0ba7e623faca970e82a65

This is a great improvement, my review can be addressed later.

File: contrib/check-semver.sh
Comment (storopoli) [reply contrib/check-semver.sh:114]: Yeah that was my initial implementation.

File: contrib/check-semver.sh
Comment (storopoli) [reply contrib/check-semver.sh:183]: This is if all cargo semver-checks got a `FAIL`.

**Review by tcharding — APPROVED** [review]
ACK eeae225cfca149d669b0ba7e623faca970e82a65

File: contrib/check-semver.sh
Comment (Kixunil) [reply contrib/check-semver.sh:183]: Oh, doing it in a loop would create the comment multiple times, right?

**Comment by tcharding** [comment]
> It reproduces the current behavior of cargo-public-api:

Damn, I missed this during review. This claim is not correct. Its missing "all features" for `hashes`, `units`, and `io`.

``bash
# Uses `CARGO` to generate API files in the specified crate.
#
# Files:
#
# - no-features.txt
# - alloc-only.txt
# - all-features.txt
generate_api_files() {
    local crate=$1
    pushd "$REPO_DIR/$crate" > /dev/null

    run_cargo --no-default-features | $SORT | uniq > "$API_DIR/$crate/no-features.txt"
    run_cargo --no-default-features --features=alloc | $SORT | uniq > "$API_DIR/$crate/alloc-only.txt"
    run_cargo_all_features | $SORT | uniq > "$API_DIR/$crate/all-features.txt"

    popd > /dev/null
}
```

Was there a reason for this @storopoli?

**Comment by storopoli** [comment]
@tcharding no, my intention was to mimic what we did with `cargo-public-api`.
From my reading of the script we only checked:

- `bitcoin`:
  - `all-features`
  - `no-default-features`
- `base58ck`:
  - `all-features`
  - `no-default-features`
- `bitcoin_hashes`:
  - `no-default-features`
  - `features alloc`
- `bitcoin-units`:
  - `no-default-features`
  - `features alloc`
- `bitcoin-io`:
  - `no-default-features`
  - `features alloc`

But yeah, I need to add the `all-features` to `bitcoin_hashes` `bitcoin-units` and `bitcoin-io`.
Thanks for pointing that out.

**Comment by tcharding** [comment]
My list is not correct, the `generate_api_files` function I posted above was called for `hashes`, `units`, and `io`.

**Pull request merged**
